lang/funcs/collection.go

```diff
@@ -461,6 +461,53 @@ var MatchkeysFunc = function.New(&function.Spec{
 	},
 })
 
+// SumFunc constructs a function that returns the sum of all
+// numbers provided in a list
+var SumFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.DynamicPseudoType,
+		},
+	},
+	Type: function.StaticReturnType(cty.Number),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		if !args[0].CanIterateElements() {
+			return cty.NilVal, function.NewArgErrorf(0, "cannot sum noniterable")
+		}
+
+		if args[0].LengthInt() == 0 { // Easy path
+			return cty.NilVal, function.NewArgErrorf(0, "cannot sum an empty list")
+		}
+
+		arg := args[0].AsValueSlice()
+		ty := args[0].Type()
+
+		var i float64
+		var s float64
+
+		if !ty.IsListType() && !ty.IsSetType() && !ty.IsTupleType() {
+			return cty.NilVal, function.NewArgErrorf(0, fmt.Sprintf("argument must be list, set, or tuple. Received %s", ty.FriendlyName()))
+		}
+
+		if !args[0].IsKnown() {
+			return cty.UnknownVal(cty.Number), nil
+		}
+
+		for _, v := range arg {
+
+			if err := gocty.FromCtyValue(v, &i); err != nil {
+				return cty.UnknownVal(cty.String), function.NewArgErrorf(0, "argument must be list, set, or tuple of number values.")
```

Comment:
I think this should be `cty.UnknownVal(cty.Number)` (as on line 495 above).

Also, please remove the period from the end of the error string—the error UI adds a period, which results in a double-period:

```
$ tf apply

Error: Invalid function argument

  on main.tf line 2, in output "result":
   2:   value = sum(["a"])

Invalid value for "list" parameter: argument must be list, set, or tuple of
number values..
```